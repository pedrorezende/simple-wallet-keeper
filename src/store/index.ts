import { combineReducers, configureStore } from "@reduxjs/toolkit";
import keyringReducer from "./keyring";
import storage from "redux-persist/lib/storage";
import { persistReducer, persistStore } from "redux-persist";

const persistConfig = {
  key: "keyring-app",
  storage,
  version: 1,
  whitelist: ["keyring"],
};

const rootReducer = combineReducers({
  keyring: keyringReducer,
});

const persistedReducer = persistReducer(persistConfig, rootReducer);
export const store = configureStore({
  reducer: persistedReducer,
});

export const persistor = persistStore(store);
export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;
