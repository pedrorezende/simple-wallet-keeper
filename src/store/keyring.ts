import { createAsyncThunk, createSlice } from "@reduxjs/toolkit";
import type { PayloadAction } from "@reduxjs/toolkit";
import type { RootState } from "store";
import { WalletService } from "services/WalletService";

// Define a type for the slice state
interface KeyringState {
  version: number;
  initialWalletAddress?: string;
  wallets?: {
    [publicKey: string]: string;
  };
}

type CreateWalletParams = {
  password: string;
  initialWallet: boolean;
};

// Define the initial state using that type
const initialState: KeyringState = {
  version: 1,
};

export const createWallet = createAsyncThunk(
  "keyring/createWallet",
  async ({ password, initialWallet = false }: CreateWalletParams) => {
    return {
      wallet: await WalletService.create(password),
      initialWallet,
    };
  }
);

export const keyringSlice = createSlice({
  name: "keyring",
  initialState,
  reducers: {
    removeWallet: (state, action: PayloadAction<string>) => {
      if (state.wallets) {
        delete state.wallets[action.payload];
      }
    },
  },
  extraReducers: (builder) => {
    builder.addCase(createWallet.fulfilled, (state, action) => {
      const [address, encryptedWallet] = action.payload.wallet;
      if (!state.wallets) {
        state.wallets = {};
      }

      if (action.payload.initialWallet) {
        state.initialWalletAddress = address;
      }

      state.wallets[address] = encryptedWallet;
    });
  },
});

export const selectKeyringWallets = (state: RootState) => state.keyring.wallets;
export const selectInitialWalletAddress = (state: RootState) =>
  state.keyring.initialWalletAddress;

export const { removeWallet } = keyringSlice.actions;
export default keyringSlice.reducer;
