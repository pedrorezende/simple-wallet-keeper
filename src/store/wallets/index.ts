import { createSlice } from "@reduxjs/toolkit";
import type { PayloadAction } from "@reduxjs/toolkit";
import type { RootState } from "store";

// Define a type for the slice state
interface WalletState {
  version: number;
}

// Define the initial state using that type
const initialState: WalletState = {
  version: 1,
};

export const walletsSlice = createSlice({
  name: "wallets",
  initialState,
  reducers: {
    create: (state, action: PayloadAction<number>) => {
      state.version = 1;
    },
  },
});

export const { create } = walletsSlice.actions;

// Other code such as selectors can use the imported `RootState` type
export const selectWallet = (state: RootState) => state.wallets.value;

export default walletsSlice.reducer;
