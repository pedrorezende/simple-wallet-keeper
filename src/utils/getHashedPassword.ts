import shajs from "sha.js";

export const getHashedPassword = (password: string) =>
  shajs("sha256").update(password).digest("hex");
