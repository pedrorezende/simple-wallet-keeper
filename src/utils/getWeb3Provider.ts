import { ethers } from "ethers";

const provider = new ethers.providers.JsonRpcProvider(
  process.env.NEXT_PUBLIC_PROVIDER_RPC
);

export { provider };
