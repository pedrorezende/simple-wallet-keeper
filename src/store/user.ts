import { createSlice } from "@reduxjs/toolkit";
import type { PayloadAction } from "@reduxjs/toolkit";
import { RootState } from "store";

// Define a type for the slice state
interface UserState {
  hashedPassword: string;
}

// Define the initial state using that type
const initialState: UserState = {
  hashedPassword: "",
};

export const userSlice = createSlice({
  name: "user",
  initialState,
  reducers: {
    storeHashedPassword: (state, action: PayloadAction<string>) => {
      state.hashedPassword = action.payload;
    },
  },
});

export const selectPasswordHash = (state: RootState) =>
  state.user.hashedPassword;

export const { storeHashedPassword } = userSlice.actions;
export default userSlice.reducer;
