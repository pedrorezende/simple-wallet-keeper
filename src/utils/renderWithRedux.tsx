import React from "react";
import { Provider } from "react-redux";
import { store } from "store";

export const renderWithRedux = (component: React.ReactNode) => {
  return <Provider store={store}>{component}</Provider>;
};
