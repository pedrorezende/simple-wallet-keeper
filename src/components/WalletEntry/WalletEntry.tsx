import { PrimaryKeyPanel } from "components/PrimaryKeyPanel";
import { useWalletFunds } from "hooks/useWalletFunds";
import useTranslation from "next-translate/useTranslation";
import React, { useState } from "react";
import { useAppDispatch } from "store/hooks";
import { removeWallet } from "store/keyring";

export interface WalletEntryProps {
  encryptedWallet: string;
  address: string;
  canDelete: boolean;
}

export function WalletEntry({
  encryptedWallet,
  address,
  canDelete,
}: WalletEntryProps) {
  const { t } = useTranslation();
  const dispatch = useAppDispatch();
  const balance = useWalletFunds({ walletAddress: address });
  const [displayPrimaryKey, setDisplayPrimaryKey] = useState(false);

  const onRemove = () => {
    if (confirm(t("common:deletion_prompt"))) {
      dispatch(removeWallet(address));
    }
  };

  return (
    <li className="flex justify-between px-6 py-3 font-light text-gray-500 bg-gray-100 rounded-md shadow-sm text-md hover:bg-gray-50">
      <div>
        <div>{address}</div>
        {displayPrimaryKey && (
          <PrimaryKeyPanel
            onClose={() => setDisplayPrimaryKey(false)}
            encryptedWallet={encryptedWallet}
          />
        )}
        <div className="flex gap-3 mt-2 text-xs">
          {!displayPrimaryKey && (
            <button
              role="button"
              onClick={() => setDisplayPrimaryKey(true)}
              className="text-blue-600 uppercase"
            >
              {t("common:view_primary_key")}
            </button>
          )}

          {displayPrimaryKey && (
            <button
              role="button"
              onClick={() => setDisplayPrimaryKey(false)}
              className="text-blue-600 uppercase"
            >
              {t("common:hide_primary_key")}
            </button>
          )}

          {canDelete && (
            <button
              role="button"
              className="text-red-500 uppercase"
              onClick={onRemove}
            >
              {t("common:delete_wallet")}
            </button>
          )}
        </div>
      </div>
      <div>{balance}</div>
    </li>
  );
}
