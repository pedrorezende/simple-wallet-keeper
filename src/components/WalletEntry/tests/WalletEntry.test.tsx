import React from "react";
import { WalletEntry, WalletEntryProps } from "../WalletEntry";
import { fireEvent, render, screen } from "@testing-library/react";
import { renderWithRedux } from "utils/renderWithRedux";
import { removeWallet } from "store/keyring";

const walletFunds = "100 tBNB";
jest.mock("hooks/useWalletFunds", () => ({
  useWalletFunds: () => walletFunds,
}));

const removeWalletMock = jest.fn();
jest.mock("store/hooks", () => ({
  useAppDispatch: () => {
    return removeWalletMock;
  },
}));

describe("<WalletEntry />", () => {
  let props: WalletEntryProps;

  beforeEach(() => {
    props = {
      address: "wallet-address",
      canDelete: true,
      encryptedWallet: "...",
    };

    window.confirm = jest.fn().mockImplementation(() => true);
  });

  it("should display wallet address", () => {
    render(renderWithRedux(<WalletEntry {...props} />));
    expect(screen.getByText(props.address)).toBeInTheDocument();
    expect(screen).toMatchSnapshot();
  });

  it("should not display remove button", () => {
    render(renderWithRedux(<WalletEntry {...props} canDelete={false} />));
    expect(screen.queryByText("common:delete_wallet")).not.toBeInTheDocument();
    expect(screen).toMatchSnapshot();
  });

  it("should display balance and handle empty values", () => {
    render(renderWithRedux(<WalletEntry {...props} />));
    expect(screen.getByText(walletFunds)).toBeInTheDocument();
    expect(screen).toMatchSnapshot();
  });

  it("should call wallet removal", () => {
    render(renderWithRedux(<WalletEntry {...props} />));
    const deleteButton = screen.getByText("common:delete_wallet");
    fireEvent.click(deleteButton);
    expect(removeWalletMock).toHaveBeenCalledWith(removeWallet(props.address));
    expect(screen).toMatchSnapshot();
  });

  it("should display primary key panel", async () => {
    render(renderWithRedux(<WalletEntry {...props} />));
    const viewButton = screen.getByText("common:view_primary_key");
    viewButton.click();
    expect(
      await screen.findByText("private_key_instructions", { exact: false })
    );
    expect(screen).toMatchSnapshot();
  });
});
