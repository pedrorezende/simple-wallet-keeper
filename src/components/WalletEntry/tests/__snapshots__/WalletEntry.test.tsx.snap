// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<WalletEntry /> should call wallet removal 1`] = `
{
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "logTestingPlaygroundURL": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
}
`;

exports[`<WalletEntry /> should display balance and handle empty values 1`] = `
{
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "logTestingPlaygroundURL": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
}
`;

exports[`<WalletEntry /> should display primary key panel 1`] = `
{
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "logTestingPlaygroundURL": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
}
`;

exports[`<WalletEntry /> should display wallet address 1`] = `
{
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "logTestingPlaygroundURL": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
}
`;

exports[`<WalletEntry /> should not display remove button 1`] = `
{
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "logTestingPlaygroundURL": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
}
`;
